src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Areas/Identity/RevalidatingAuthenticationStateProvider.cs

```diff
@@ -1,98 +1,59 @@
-using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.Server;
 using Microsoft.AspNetCore.Identity;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
 
 namespace BlazorServerWeb_CSharp.Areas.Identity
 {
-    /// <summary>
-    /// An <see cref="AuthenticationStateProvider"/> service that revalidates the
-    /// authentication state at regular intervals. If a signed-in user's security
-    /// stamp changes, this revalidation mechanism will sign the user out.
-    /// </summary>
-    /// <typeparam name="TUser">The type encapsulating a user.</typeparam>
-    public class RevalidatingAuthenticationStateProvider<TUser>
-        : AuthenticationStateProvider, IDisposable where TUser : class
+    public class RevalidatingIdentityAuthenticationStateProvider<TUser>
+        : RevalidatingServerAuthenticationStateProvider where TUser : class
     {
         private readonly static TimeSpan RevalidationInterval = TimeSpan.FromMinutes(30);
-
-        private readonly CancellationTokenSource _loopCancellationTokenSource = new CancellationTokenSource();
         private readonly IServiceScopeFactory _scopeFactory;
-        private readonly ILogger _logger;
-        private Task<AuthenticationState> _currentAuthenticationStateTask;
+        private readonly IdentityOptions _options;
 
-        public RevalidatingAuthenticationStateProvider(
+        public RevalidatingIdentityAuthenticationStateProvider(
+            ILoggerFactory loggerFactory,
             IServiceScopeFactory scopeFactory,
-            SignInManager<TUser> circuitScopeSignInManager,
-            ILogger<RevalidatingAuthenticationStateProvider<TUser>> logger)
+            IOptions<IdentityOptions> optionsAccessor)
+            : base(loggerFactory, RevalidationInterval)
         {
-            var initialUser = circuitScopeSignInManager.Context.User;
-            _currentAuthenticationStateTask = Task.FromResult(new AuthenticationState(initialUser));
             _scopeFactory = scopeFactory;
-            _logger = logger;
-
-            if (initialUser.Identity.IsAuthenticated)
-            {
-                _ = RevalidationLoop();
-            }
+            _options = optionsAccessor?.Value ?? new IdentityOptions();
         }
 
-        public override Task<AuthenticationState> GetAuthenticationStateAsync()
-            => _currentAuthenticationStateTask;
-
-        private async Task RevalidationLoop()
+        protected override async Task<bool> ValidateAuthenticationStateAsync(AuthenticationState authenticationState)
         {
-            var cancellationToken = _loopCancellationTokenSource.Token;
-
-            while (!cancellationToken.IsCancellationRequested)
+            // Get the user manager from a new scope to ensure it fetches fresh data
+            using (var scope = _scopeFactory.CreateScope())
```

Comment:
This needs to be `await using` https://github.com/aspnet/AspNetCore/issues/12918

Reply:
Not sure it will be that easy since `IServiceScope` doesn't derive from `IAsyncDisposable` (only the MSFT DI implementation does). You'll likely have to do some `as` casting 😕

Reply:
That's quite a nasty mess given that this is going into user code in the template.

It feels like `await using` should be able to work with `IDisposable` too. If this can be implemented in user code without there being anything opinionated about it, the compiler should be able to codegen the same.

Reply:
I've handled it in what seems to be [the most beginner-friendly manner](https://github.com/aspnet/AspNetCore/pull/12909/files#diff-8a43f02a3a6b89391c3f42a90fdefa82R36), but it's still verbose and lame 😞